File: app/src/main/kotlin/tech/mmarca/openvitals/data/repository/contract/NutritionRepository.kt
package tech.mmarca.openvitals.data.repository.contract

import java.time.LocalDate
import tech.mmarca.openvitals.core.period.PeriodLoadQuery
import tech.mmarca.openvitals.domain.model.DailyMacros
import tech.mmarca.openvitals.domain.model.NutritionEntry
import tech.mmarca.openvitals.domain.model.NutritionWriteRequest
import tech.mmarca.openvitals.domain.model.RefreshMode
import tech.mmarca.openvitals.domain.query.NutritionPeriodData

interface NutritionRepository {
    val nutritionWritePermissions: Set<String>

    suspend fun loadNutritionPeriod(
        query: PeriodLoadQuery,
        refreshMode: RefreshMode = RefreshMode.NORMAL,
    ): NutritionPeriodData

    suspend fun loadDailyMacros(start: LocalDate, end: LocalDate): List<DailyMacros>

    suspend fun loadNutritionEntries(start: LocalDate, end: LocalDate): List<NutritionEntry>

    suspend fun hasNutritionWritePermission(): Boolean

    suspend fun writeCarbsEntry(request: NutritionWriteRequest): String

    suspend fun writeNutritionEntry(request: NutritionWriteRequest): String

    suspend fun deleteNutritionEntry(id: String)
}

File: lib/data/repository/contract/nutrition_repository.dart
import '../../../core/period/period_load_query.dart';
import '../../../core/result/result.dart';
import '../../../core/time/local_date.dart';
import '../../../domain/model/nutrition_models.dart';
import '../../../domain/model/refresh_mode.dart';
import '../../../domain/query/nutrition_period_data.dart';

/// Port of the Kotlin `NutritionRepository` contract.
///
/// Fallible operations return [Result]; the synchronous probe
/// ([nutritionWritePermissions]) reads cached state and cannot fail, so it
/// stays bare.
abstract interface class NutritionRepository {
  Set<String> get nutritionWritePermissions;

  Future<Result<NutritionPeriodData>> loadNutritionPeriod(
    PeriodLoadQuery query, {
    RefreshMode refreshMode = RefreshMode.normal,
  });

  Future<Result<List<DailyMacros>>> loadDailyMacros(
    LocalDate start,
    LocalDate end,
  );

  Future<Result<List<NutritionEntry>>> loadNutritionEntries(
    LocalDate start,
    LocalDate end,
  );

  Future<Result<bool>> hasNutritionWritePermission();

  Future<Result<String>> writeCarbsEntry(NutritionWriteRequest request);

  Future<Result<String>> writeNutritionEntry(NutritionWriteRequest request);

  Future<Result<void>> deleteNutritionEntry(String id);
}
